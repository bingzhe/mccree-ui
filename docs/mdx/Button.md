---
title: Button 按钮
order: 0
mobile: true
group:
    title: 基础组件
    order: 0
---

  <!-- path: base -->

## Butt

Demo:

## Foo

<!-- Demo: -->

```tsx
import React from "react";
import { Button } from "mccree-ui/components";
import "@mccree-ui/components/Button/style/index"

const a = () => {
    return <Button>default</Button>;
};

export default a;
```

<code src="../demo/test.tsx"></code>
