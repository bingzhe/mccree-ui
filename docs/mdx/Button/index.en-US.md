---
title: Button
mobile: true
nav:
    path: /components
---

  <!-- path: base -->

## Button  en-us

```tsx
import React from "react";
import { Button } from "mccree-ui/components";
// import "@mccree-ui/components/Button/style/index";

const a = () => {
    return <Button>default</Button>;
};

export default a;
```
