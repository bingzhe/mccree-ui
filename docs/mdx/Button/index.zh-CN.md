---
title: Button 按钮
mobile: true
nav:
    path: /components
---

## Button

```tsx
import React from "react";
import { Button } from "@mccree-ui/components";

const a = () => {
    return <Button>default</Button>;
};

export default a;
```

<code src="../../demo/test.tsx"></code>

// import "@mccree-ui/components/Button/style/index";
