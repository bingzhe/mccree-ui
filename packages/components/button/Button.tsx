import React, { useContext, forwardRef, useRef } from "react";
import { cs } from "@mccree-ui/util";
// import ButtonGroup from "./ButtonGroup";
import { ConfigContext } from "../ConfigProvider";
import type { ButtonProps } from "./Button.type";
import { useMergeProps } from "@mccree-ui/hooks";
import { IconLoading } from "@mccree-ui/icons";

// interface ButtonFC extends React.FC<ButtonProps> {
//     Group: typeof ButtonGroup;
//     IconButton: typeof IconButton;
// }

const defaultProps: ButtonProps = {
    variant: "default",
    theme: "default",
    shape: "square"
};

const InternalButton: React.ForwardRefRenderFunction<any, ButtonProps> = (baseProps, ref) => {
    const { getPrefixCls, size: ctxSize, componentConfig } = useContext(ConfigContext);
    const props = useMergeProps(baseProps, defaultProps, componentConfig?.Button);

    const {
        className,
        children,
        variant,
        theme,
        size,
        shape,
        disabled,
        loading,
        block,
        ghost,
        icon,
        iconOnly,
        href,
        anchorProps,
        onClick,
        ...restProps
    } = props;

    const iconNode = loading ? <IconLoading /> : icon;

    const innerButtonRef = useRef();
    const buttonRef = ref || innerButtonRef;

    const handleClick: React.MouseEventHandler<HTMLElement> = (e) => {
        if (loading) return;
        onClick?.(e);
    };

    //TODO IconOnly
    const prefixCls = getPrefixCls("btn");
    const classes = cs(
        className,
        prefixCls,
        `${prefixCls}-variant-${variant}`,
        `${prefixCls}-theme-${theme}`,
        `${prefixCls}-size-${size || ctxSize}`,
        {
            [`${prefixCls}-shape-${shape}`]: shape !== "square",
            // [`${prefixCls}-size-${size || ctxSize}`]: (size || ctxSize) !== "default",
            [`${prefixCls}-loading`]: loading,
            [`${prefixCls}-ghost`]: ghost,
            [`${prefixCls}-disabled`]: disabled,
            [`${prefixCls}-icon`]: iconNode,
            [`${prefixCls}-icon-only`]: iconOnly || (!children && children !== 0 && iconNode),
            [`${prefixCls}-block`]: block
        }
    );

    const InnerContent = (
        <>
            {iconNode}
            {children ? <span>children</span> : ""}
        </>
    );

    if (href) {
        const __anchorProps = { ...anchorProps };

        if (disabled) {
            delete __anchorProps.href;
        } else {
            __anchorProps.href = href;
        }

        return (
            <a
                ref={buttonRef}
                {...restProps}
                {...__anchorProps}
                className={classes}
                onClick={handleClick}
            >
                {InnerContent}
            </a>
        );
    }

    return (
        <button
            ref={buttonRef}
            {...restProps}
            className={classes}
            disabled={disabled}
            onClick={handleClick}
        >
            {InnerContent}
        </button>
    );
};

const ForwardRefButton = forwardRef(InternalButton);
const Button = ForwardRefButton as typeof ForwardRefButton & {
    // Group: typeof ButtonGroup;
    __MR_BUTTON: boolean;
};

// Button.Group = ButtonGroup;
Button.__MR_BUTTON = true;
Button.displayName = "Button";

export default Button;
