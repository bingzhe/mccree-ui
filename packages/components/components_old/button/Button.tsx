import * as React from "react";
import classNames from "classnames";
import PropTypes from "prop-types";

import Ripple from "../ripple";
import ButtonGroup from "./ButtonGroup";
import IconButton from "./IconButton";
import { ConfigContext } from "../config-provider";
import SizeContext, { SizeType } from "../config-provider/SizeContext";

import { omit } from "../utils/omit";
import { tuple } from "../utils/type";
import Icon from "../icon/index";

const ButtonVariants = tuple("contain", "outline", "text");
export type ButtonVariant = typeof ButtonVariants[number];
const ButtonTypes = tuple("primary", "secondary", "success", "warning", "error", "info");
export type ButtonType = typeof ButtonTypes[number];
const ButtonColors = tuple("primary", "secondary", "success", "warning", "error", "info");
export type ButtonColor = typeof ButtonColors[number];
const ButtonHTMLTypes = tuple("submit", "button", "reset");
export type ButtonHTMLType = typeof ButtonHTMLTypes[number];
const ButtonShapes = tuple("circle", "round");
export type ButtonShape = typeof ButtonShapes[number];

export interface BaseButtonProps {
    type?: ButtonType;
    color?: ButtonColor;
    variant?: ButtonVariant;
    shape?: ButtonShape;
    size: SizeType;
    disabled?: boolean;
    className?: string;
    prefixCls?: string;
    loading?: boolean | { delay?: number };
    block?: boolean;
    startIcon?: React.ReactNode;
    endIcon?: React.ReactNode;
    children?: React.ReactNode;
    centerRipple?: boolean;
}

export type AnchorButtonProps = {
    href: string;
    target?: string;
    onClick: React.MouseEventHandler<HTMLElement>;
} & BaseButtonProps &
    Omit<React.AnchorHTMLAttributes<any>, "type" | "onClick">;

export type NativeButtonProps = {
    htmlType?: ButtonHTMLType;
    onClick?: React.MouseEventHandler<HTMLElement>;
} & BaseButtonProps &
    Omit<React.ButtonHTMLAttributes<any>, "type" | "onClick">;

export type ButtonProps = Partial<AnchorButtonProps | NativeButtonProps>;

interface ButtonFC extends React.FC<ButtonProps> {
    Group: typeof ButtonGroup;
    IconButton: typeof IconButton;
}

const Button: ButtonFC = (props) => {
    const {
        type,
        color,
        shape,
        size: sizeProp,
        className,
        children,
        block,
        variant,
        startIcon: startIconProp,
        endIcon: endIconProp,
        loading: loadingProp,
        onClick,
        centerRipple,
        ...restProps
    } = props;

    const { disabled } = restProps;

    const [loading, setLoading] = React.useState(loadingProp);
    const { getPrefixCls } = React.useContext(ConfigContext);
    const sizeContext = React.useContext(SizeContext);
    const delayTimeout = React.useRef<number>();

    React.useEffect(() => {
        if (loadingProp && typeof loadingProp !== "boolean") {
            clearTimeout(delayTimeout.current);
        }
        if (loadingProp && typeof loadingProp !== "boolean" && loadingProp.delay) {
            delayTimeout.current = window.setTimeout(() => {
                setLoading(loadingProp);
            }, loadingProp.delay);
        } else if (loadingProp !== loading) {
            setLoading(loadingProp);
        }
    }, [loading, loadingProp]);

    const handleClick: React.MouseEventHandler<HTMLElement> = (e) => {
        if (loading) return;
        onClick?.(e);
    };

    const prefixCls = getPrefixCls("btn");
    const rootPrefixCls = getPrefixCls("btn-root");

    const rootClasses = classNames(rootPrefixCls, className);

    let rippleColor: string | undefined;
    let rippleBorderRadius = "4px";
    let sizeCls = "";

    switch (sizeProp || sizeContext) {
        case "large":
            sizeCls = "lg";
            rippleBorderRadius = shape === "round" ? "40px" : "4px";
            break;
        case "small":
            sizeCls = "sm";
            rippleBorderRadius = shape === "round" ? "24px" : "2px";
            break;
        default:
            rippleBorderRadius = shape === "round" ? "32px" : "4px";
            break;
    }

    if (variant === "outline" || variant === "text") {
        rippleColor = color;
    } else if (variant === "contain") {
        if (!color) {
            rippleColor = "#000";
        } else {
            rippleColor = "#FFF";
        }
    }
    let rippleStyle: React.CSSProperties = { borderRadius: rippleBorderRadius };

    const classes = classNames(prefixCls, {
        [`${prefixCls}-${color}`]: color,
        [`${prefixCls}-${variant}`]: variant,
        [`${prefixCls}-${shape}`]: shape,
        [`${prefixCls}-${sizeCls}`]: sizeCls,
        [`${prefixCls}-loading`]: loading,
        [`${prefixCls}-icon`]: startIconProp || endIconProp,
        [`${prefixCls}-block`]: block
    });

    const loadingIcon = loading && (
        <span className={`${prefixCls}-icon-loading`}>
            <Icon name="loading" spin />
        </span>
    );
    const startIcon = !loading && startIconProp && (
        <span
            className={classNames(`${prefixCls}-prefix-icon`, {
                [`${prefixCls}-icon-${sizeCls}`]: sizeCls
            })}
        >
            {startIconProp}
        </span>
    );
    const endIcon = !loading && endIconProp && (
        <span
            className={classNames(`${prefixCls}-suffix-icon`, {
                [`${prefixCls}-icon-${sizeCls}`]: sizeCls
            })}
        >
            {endIconProp}
        </span>
    );

    const linkButtonRestProps = omit(restProps as AnchorButtonProps, ["loading"]);

    if (linkButtonRestProps.href !== undefined) {
        return (
            <a {...linkButtonRestProps} className={classes} onClick={handleClick}>
                {children}
            </a>
        );
    }

    const { htmlType, ...otherProps } = restProps as NativeButtonProps;

    const buttonNode = (
        <Ripple
            block={block}
            style={rippleStyle}
            center={centerRipple}
            className={rootClasses}
            color={rippleColor}
            disableRipple={disabled}
        >
            <button
                {...(omit(otherProps, ["loading"]) as NativeButtonProps)}
                type={htmlType}
                className={classes}
                onClick={handleClick}
            >
                {/* disabled={disabled} */}
                {loadingIcon}
                {startIcon}
                {children}
                {endIcon}
            </button>
        </Ripple>
    );

    return buttonNode;
};

Button.defaultProps = {
    variant: "contain",
    size: "medium",
    disabled: false,
    loading: false,
    block: false,
    centerRipple: false
};

Button.Group = ButtonGroup;
Button.IconButton = IconButton;

Button.propTypes = {
    disabled: PropTypes.bool,
    size: PropTypes.oneOf(["small", "medium", "large"])
};

export default Button;
