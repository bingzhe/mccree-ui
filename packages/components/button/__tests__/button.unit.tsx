import * as React from "react";
import { render, mount } from "enzyme";
import toJson from "enzyme-to-json";

import Button from "../Button";

describe("button", () => {
    it("render <Button/> component", () => {
        const wrapper = render(
            <>
                <Button>DEFAULT</Button>
                <Button type="primary">PRIMARY</Button>
                <Button type="secondary">SECONDARY</Button>
                <Button type="success" >SUCCESS</Button>
                <Button type="error" >ERROE</Button>
                <Button type="warning" >WARNING</Button>
                <Button type="info" >INFO</Button>
                <Button disabled>DISABLED</Button>
                <Button target="block" href="http://bingzhe.github.io/" type="primary">LINK</Button>

                <Button type="primary" loading>loading</Button>
                <Button type="primary" size="large">大</Button>
                <Button type="primary" size="medium">中</Button>
                <Button type="primary" size="small">小</Button>

                <Button type="primary" block>loading</Button>
                <Button type="primary" shape="circle">circle</Button>
                <Button type="primary" shape="round">loading</Button>
            </>
        );


        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("should find mccree-btn classnames", () => {
        const wrapper = mount(
            <>
                <Button>DEFAULT</Button>
                <Button type="primary">PRIMARY</Button>
                <Button type="secondary">SECONDARY</Button>
                <Button type="success" >SUCCESS</Button>
                <Button type="error" >ERROE</Button>
                <Button type="warning" >WARNING</Button>
                <Button type="info" >INFO</Button>
            </>
        );

        expect(wrapper.find(".mccree-btn").length).toBe(7);
    });

    it("should find type classnames", () => {
        const wrapper = mount(
            <>
                <Button>DEFAULT</Button>
                <Button type="primary">PRIMARY</Button>
                <Button type="secondary">SECONDARY</Button>
                <Button type="success" >SUCCESS</Button>
                <Button type="error" >ERROE</Button>
                <Button type="warning" >WARNING</Button>
                <Button type="info" >INFO</Button>
            </>
        );

        expect(wrapper.find(".mccree-btn-primary").length).toBe(1);
        expect(wrapper.find(".mccree-btn-secondary").length).toBe(1);
        expect(wrapper.find(".mccree-btn-success").length).toBe(1);
        expect(wrapper.find(".mccree-btn-error").length).toBe(1);
        expect(wrapper.find(".mccree-btn-warning").length).toBe(1);
        expect(wrapper.find(".mccree-btn-info").length).toBe(1);
    });

    it("should find variant classnames", () => {
        const wrapper = mount(
            <>
                <Button type="primary">PRIMARY</Button>
                <Button type="primary" variant="outline">PRIMARY</Button>
                <Button type="primary" variant="text">PRIMARY</Button>
            </>
        );

        expect(wrapper.find(".mccree-btn-contain").length).toBe(1);
        expect(wrapper.find(".mccree-btn-outline").length).toBe(1);
        expect(wrapper.find(".mccree-btn-text").length).toBe(1);
    });


    it("should can trigger click event", () => {
        const onClick = jest.fn();
        const wrapper = mount(
            <Button type="primary" onClick={onClick}>PRIMARY</Button>
        );
        wrapper.simulate("click");
        expect(onClick).toHaveBeenCalled();
    });

    it("should can not trigger click event when button is disabled", () => {
        const onClick = jest.fn();
        const wrapper = mount(
            <Button disabled type="primary" onClick={onClick}>PRIMARY</Button>
        );
        wrapper.simulate("click");
        expect(onClick).not.toHaveBeenCalled();
    });

    it("should can not trigger click event when button is loading", () => {
        const onClick = jest.fn();
        const wrapper = mount(
            <Button loading type="primary" onClick={onClick}>PRIMARY</Button>
        );
        wrapper.simulate("click");
        expect(onClick).not.toHaveBeenCalled();
    });
});