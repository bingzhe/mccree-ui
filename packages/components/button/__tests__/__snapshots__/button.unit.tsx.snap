// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button render <Button/> component 1`] = `
Array [
  <button
    class="mccree-btn mccree-btn-contain"
  >
    DEFAULT
    <span
      class="mccree-ripple-root"
    />
  </button>,
  <button
    class="mccree-btn mccree-btn-primary mccree-btn-contain"
  >
    PRIMARY
    <span
      class="mccree-ripple-root"
    />
  </button>,
  <button
    class="mccree-btn mccree-btn-secondary mccree-btn-contain"
  >
    SECONDARY
    <span
      class="mccree-ripple-root"
    />
  </button>,
  <button
    class="mccree-btn mccree-btn-success mccree-btn-contain"
  >
    SUCCESS
    <span
      class="mccree-ripple-root"
    />
  </button>,
  <button
    class="mccree-btn mccree-btn-error mccree-btn-contain"
  >
    ERROE
    <span
      class="mccree-ripple-root"
    />
  </button>,
  <button
    class="mccree-btn mccree-btn-warning mccree-btn-contain"
  >
    WARNING
    <span
      class="mccree-ripple-root"
    />
  </button>,
  <button
    class="mccree-btn mccree-btn-info mccree-btn-contain"
  >
    INFO
    <span
      class="mccree-ripple-root"
    />
  </button>,
  <button
    class="mccree-btn mccree-btn-contain"
    disabled=""
  >
    DISABLED
    <span
      class="mccree-ripple-root"
    />
  </button>,
  <a
    class="mccree-btn mccree-btn-primary mccree-btn-contain"
    href="http://bingzhe.github.io/"
    target="block"
  >
    LINK
  </a>,
  <button
    class="mccree-btn mccree-btn-primary mccree-btn-contain mccree-btn-loading"
  >
    <span
      class="mccree-btn-icon-loading"
    >
      <svg
        class="mccree-icon mccree-icon-spin"
        name="loading"
      >
        <use
          href="#loading"
        />
      </svg>
    </span>
    loading
    <span
      class="mccree-ripple-root"
    />
  </button>,
  <button
    class="mccree-btn mccree-btn-primary mccree-btn-contain mccree-btn-lg"
  >
    大
    <span
      class="mccree-ripple-root"
    />
  </button>,
  <button
    class="mccree-btn mccree-btn-primary mccree-btn-contain"
  >
    中
    <span
      class="mccree-ripple-root"
    />
  </button>,
  <button
    class="mccree-btn mccree-btn-primary mccree-btn-contain mccree-btn-sm"
  >
    小
    <span
      class="mccree-ripple-root"
    />
  </button>,
  <button
    class="mccree-btn mccree-btn-primary mccree-btn-contain mccree-btn-block"
  >
    loading
    <span
      class="mccree-ripple-root"
    />
  </button>,
  <button
    class="mccree-btn mccree-btn-primary mccree-btn-contain mccree-btn-circle"
  >
    circle
    <span
      class="mccree-ripple-root"
    />
  </button>,
  <button
    class="mccree-btn mccree-btn-primary mccree-btn-contain mccree-btn-round"
  >
    loading
    <span
      class="mccree-ripple-root"
    />
  </button>,
]
`;
